lib/Conversion/ImportVerilog/AssertionExpr.cpp

```diff
@@ -0,0 +1,344 @@
+                                       builder.getI64IntegerAttr(0));
+      auto antecedent = builder.create<ltl::ConcatOp>(
+          loc, SmallVector<Value, 2>{lhsDelay, constOne});
+      return builder.create<ltl::ImplicationOp>(
+          loc, SmallVector<Value, 2>{antecedent, rhs});
+    }
+    case BinaryAssertionOperator::OverlappedFollowedBy: {
+      auto notRhs = builder.create<ltl::NotOp>(loc, rhs);
+      auto implication = builder.create<ltl::ImplicationOp>(
+          loc, SmallVector<Value, 2>{lhs, notRhs});
+      return builder.create<ltl::NotOp>(loc, implication);
+    }
+    case BinaryAssertionOperator::NonOverlappedFollowedBy: {
+      auto constOne =
+          builder.create<hw::ConstantOp>(loc, builder.getI1Type(), 1);
+      auto notRhs = builder.create<ltl::NotOp>(loc, rhs);
+      auto lhsDelay =
+          builder.create<ltl::DelayOp>(loc, lhs, builder.getI64IntegerAttr(1),
+                                       builder.getI64IntegerAttr(0));
+      auto antecedent = builder.create<ltl::ConcatOp>(
+          loc, SmallVector<Value, 2>{lhsDelay, constOne});
+      auto implication = builder.create<ltl::ImplicationOp>(
+          loc, SmallVector<Value, 2>{antecedent, notRhs});
+      return builder.create<ltl::NotOp>(loc, implication);
+    }
+    case BinaryAssertionOperator::SUntil:
+    case BinaryAssertionOperator::SUntilWith:
+      mlir::emitError(loc, "unsupported binary operator: ")
+          << slang::ast::toString(expr.op);
+      return {};
+    }
+    llvm_unreachable("All enum values handled in switch");
+  }
+
+  Value visit(const slang::ast::ClockingAssertionExpr &expr) {
+    auto assertionExpr = context.convertAssertionExpression(expr.expr, loc);
+    if (!assertionExpr)
+      return {};
+    auto result = context.convertLTLTimingControl(expr.clocking, assertionExpr);
+    if (!result) {
+      mlir::emitError(loc, "Failed to convert clocking assertion expression");
+      return {};
+    }
+    return result;
+  }
+
+  /// Emit an error for all other expressions.
+  template <typename T>
+  Value visit(T &&node) {
+    mlir::emitError(loc, "unsupported expression: ")
+        << slang::ast::toString(node.kind);
+    return {};
+  }
+
+  Value visitInvalid(const slang::ast::AssertionExpr &expr) {
+    mlir::emitError(loc, "invalid expression");
+    return {};
+  }
+};
+} // namespace
+
+Value Context::convertAssertionExpression(const slang::ast::AssertionExpr &expr,
+                                          Location loc) {
+  AssertionExprVisitor visitor{*this, loc};
+  return expr.visit(visitor);
+}
+// NOLINTEND(misc-no-recursion)
+
+/// Helper function to convert a value to an i1 value.
+Value Context::convertToI1(Value value) {
+  if (!value)
+    return {};
+  if (auto type = dyn_cast_or_null<moore::IntType>(value.getType()))
+    if (type.getBitSize() != 1) {
+      mlir::emitError(value.getLoc(), "expected a 1-bit integer");
+      return {};
+    }
```

Comment:
LLVM likes to use `{...}` on either all or none of the nested ifs. You should also be able to use `dyn_cast`, since `value.getType()` will always return a type. The user might also do something weird where the value isn't a Moore `IntType` at all -- should be caught by the frontend, but it might be worth emitting an error then as well:
```suggestion
  auto type = dyn_cast<moore::IntType>(value.getType());
  if (!type || type.getBitSize() != 1) {
    mlir::emitError(value.getLoc(), "expected a 1-bit integer");
    return {};
  }
```

Reply:
Thanks for the detailed explanation!